modules/apps/transfer/transfer_test.go

```diff
@@ -37,101 +37,160 @@ func (suite *TransferTestSuite) SetupTest() {
+			[]string{sdk.DefaultBondDenom, ibctesting.SecondaryDenom},
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			// setup between chainA and chainB
+			// NOTE:
+			// pathAToB.EndpointA = endpoint on chainA
+			// pathAToB.EndpointB = endpoint on chainB
+			pathAToB := ibctesting.NewTransferPath(suite.chainA, suite.chainB)
+			pathAToB.Setup()
+			traceAToB := types.NewTrace(pathAToB.EndpointB.ChannelConfig.PortID, pathAToB.EndpointB.ChannelID)
+
+			originalBalances := sdk.NewCoins()
+			for _, denom := range tc.sourceDenomsToTransfer {
+				originalBalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), suite.chainA.SenderAccount.GetAddress(), denom)
+				originalBalances = originalBalances.Add(originalBalance)
+			}
+
+			timeoutHeight := clienttypes.NewHeight(1, 110)
+
+			amount, ok := sdkmath.NewIntFromString("9223372036854775808") // 2^63 (one above int64)
+			suite.Require().True(ok)
+			originalCoins := sdk.NewCoins()
+			for _, denom := range tc.sourceDenomsToTransfer {
+				coinToSendToB := sdk.NewCoin(denom, amount)
+				originalCoins = originalCoins.Add(coinToSendToB)
+			}
+
+			// send from chainA to chainB
+			msg := types.NewMsgTransfer(pathAToB.EndpointA.ChannelConfig.PortID, pathAToB.EndpointA.ChannelID, originalCoins, suite.chainA.SenderAccount.GetAddress().String(), suite.chainB.SenderAccount.GetAddress().String(), timeoutHeight, 0, "")
+			res, err := suite.chainA.SendMsgs(msg)
+			suite.Require().NoError(err) // message committed
+
+			packet, err := ibctesting.ParsePacketFromEvents(res.Events)
+			suite.Require().NoError(err)
+
+			// relay send
+			err = pathAToB.RelayPacket(packet)
+			suite.Require().NoError(err) // relay committed
+
+			escrowAddress := types.GetEscrowAddress(packet.GetSourcePort(), packet.GetSourceChannel())
+			coinsSentFromAToB := sdk.NewCoins()
+			for _, coin := range originalCoins {
+				// check that the balance for chainA is updated
+				chainABalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), suite.chainA.SenderAccount.GetAddress(), coin.Denom)
+				suite.Require().Equal(originalBalances.AmountOf(coin.Denom).Sub(amount).Int64(), chainABalance.Amount.Int64())
+
+				// check that module account escrow address has locked the tokens
+				chainAEscrowBalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), escrowAddress, coin.Denom)
+				suite.Require().Equal(coin, chainAEscrowBalance)
+
+				// check that voucher exists on chain B
+				chainBDenom := types.NewDenom(coin.Denom, traceAToB)
+				chainBBalance := suite.chainB.GetSimApp().BankKeeper.GetBalance(suite.chainB.GetContext(), suite.chainB.SenderAccount.GetAddress(), chainBDenom.IBCDenom())
+				coinSentFromAToB := sdk.NewCoin(chainBDenom.IBCDenom(), amount)
+				suite.Require().Equal(coinSentFromAToB, chainBBalance)
+
+				coinsSentFromAToB = coinsSentFromAToB.Add(coinSentFromAToB)
```

Comment:
A suggestion for a future improvement of this test is to drop the intermediate coins tracking and keep track of the ibc token type:
- original coins sent
- ibc tokens on B (originalCoin.Denom, traceAtoB)
- ibc tokens on C (originalCoin.Denom, traceAtoB, traceBtoC)

Maybe there's another way. I've always found this test hard to write and parse, variable naming wise. Probably could add some assert funcs to reduce lines of code